feat: add duration knob

Mastersam07 opened the pull request:
Adds duration knob.

Duration should be entered in milliseconds (1000 milliseconds == 1 second).

### List of issues which are fixed by the PR
#900 

### Screenrecord


https://github.com/widgetbook/widgetbook/assets/31275429/9832ad08-643c-40b2-b10f-7fab8d44f4c8


### Checklist

- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on [Discord].

<!-- Links -->
[CLA]: https://docs.google.com/forms/d/e/1FAIpQLScuRfjUzENsLsmQgqZlGLxMKbFi7zuXoPARyXytoyQrq7ntUw/viewform
[Discord]: https://discord.com/invite/zT4AMStAJA

Mastersam07 commented:
Hi @YoussefRaafatNasry can you please take a look/review?

YoussefRaafatNasry commented on packages/widgetbook/lib/src/knobs/duration_knob.dart:
```diff
@@ -0,0 +1,33 @@
+    required super.value,
+    super.description,
+  });
+
```
Consider adding `DurationKnob.nullable` constructor, and a `durationOrNull` variant to the `KnobsBuilder`. That will be a great addition that gives a lot of flexibility. (Don't forget to add a test group for that as well).

YoussefRaafatNasry commented on packages/widgetbook/lib/src/fields/duration_field.dart:
```diff
@@ -0,0 +1,38 @@
+    super.onChanged,
+  }) : super(
+          type: FieldType.doubleInput,
+          codec: durationCodec,
```
Although I understand the reason behind extracting this to its own variable to make it more testable. I would be more leaning towards adding it to the constructor directly, just like other fields, to maintain the same code style everywhere.

_(Note that I have used the same param names, branching style, etc. from other fields)_

```suggestion
          codec: FieldCodec(
            toParam: (value) => value.inMilliseconds.toString(),
            toValue: (param) {
              return param == null
                  ? null
                  : Duration(
                      milliseconds: int.tryParse(param) ?? 0,
                    );
            },
          ),
```

YoussefRaafatNasry commented on packages/widgetbook/lib/src/fields/duration_field.dart:
```diff
@@ -0,0 +1,38 @@
+    required Duration super.initialValue,
+    super.onChanged,
+  }) : super(
+          type: FieldType.doubleInput,
```
We need to create a new `FieldType` enum member called `duration`.
This `type` is used to make Fields easily renderable by external systems (like Widgetbook Cloud)

Mastersam07 replied:
Oh was unaware of that. Thought since the duration is in ms, it would fit as a doubleInput.

Would make the change.

YoussefRaafatNasry commented on packages/widgetbook/lib/src/fields/duration_field.dart:
```diff
@@ -0,0 +1,38 @@
+  Widget toWidget(BuildContext context, String label, Duration? currentValue) {
+    return TextFormField(
+      initialValue:
+          codec.toParam(currentValue ?? initialValue ?? Duration.zero),
```
Consider extracting `Duration.zero` to a `static const` variable named `defaultDuration` _(same style as `ColorField`)_

YoussefRaafatNasry commented on packages/widgetbook/lib/src/fields/duration_field.dart:
```diff
@@ -0,0 +1,38 @@
+    return TextFormField(
+      initialValue:
+          codec.toParam(currentValue ?? initialValue ?? Duration.zero),
+      keyboardType: TextInputType.number,
```
We can also add a `suffix` of "ms" to make sure this field is visually different from other fields. This can also help users know the unit we are using, leading into a better UX.

What do you think?

```suggestion
      keyboardType: TextInputType.number,
      decoration: const InputDecoration(
        suffix: Text('ms'),
      ),
```

Mastersam07 replied:
I agree. User would be able to tell the unit. Leaves no room for confusion.

YoussefRaafatNasry reviewed (changes requested):
Really love how professional this PR is! 💙 
You have done a good job with neat implementation, adding tests, Changelog item, even a video and an example.

I've added some nitpick comments that will make the code style more matching to ours, and will improve the UX!

Can't wait to get this merged and released!

Mastersam07 commented:
> Really love how professional this PR is! 💙 You have done a good job with neat implementation, adding tests, Changelog item, even a video and an example.
> 
> I've added some nitpick comments that will make the code style more matching to ours, and will improve the UX!
> 
> Can't wait to get this merged and released!

Made the changes and added their respective test groups.

YoussefRaafatNasry commented on packages/widgetbook/lib/src/knobs/builders/knobs_builder.dart:
```diff
@@ -142,4 +143,35 @@ class KnobsBuilder {
+    return onKnobAdded(
+      DurationKnob.nullable(
+        label: label,
+        value: initialValue ?? Duration.zero,
```
If the `initialValue` is null, it should passed as-is to the Knob.

```suggestion
        value: initialValue,
```

Mastersam07 replied:
True, must have missed this.

YoussefRaafatNasry reviewed (changes requested):
One little thing I forgot to mention, but we will really appreciate if you could add the new knobs to the [docs](https://github.com/widgetbook/widgetbook/blob/main/docs/knobs/overview.mdx) as well.

Mastersam07 commented:
@YoussefRaafatNasry I have made the changes.
